docs/contributing/999-rfcs/2024-02-27-api-middleware.md

```diff
@@ -0,0 +1,291 @@
+    body: "Hello world!"
+  };
+};
+
+api.get("/hello", MyMiddleware.loggingMiddleware(handler));
+```
+
+<!--
+This section works backwards from an the end user. It is written as one or more "artifacts from the future" such as the getting started documentation (readme), user interface wireframes (or link to them), press release, changelog entry, etc.
+-->
+
+## Requirements
+
+- Write new API middleware in Wing
+- Import API middleware from Wing libraries
+- Apply middleware to individual paths
+- Apply middleware to all paths
+
+<!--
+This section is a "shopping list" of requirements for this feature.
+
+We try to start by identifying the use cases that are expected to be addressed by this RFC.
+Ideally they should not inform the design or implementation but rather state the problems/pains/results
+that our users expect to achieve with this RFC.
+
+The requirements are the "contract" of the feature you're developing - "what does it do?" (as opposed to "how does it do it" - the implementation). The requirements usually specify use cases as well as edge case scenarios and the desired behavior of the software described.
+
+NOTES:
+* It is highly recommended to split functional and non-functional requirements.
+* Requirements should be prioritized P0 (must), P1 (nice to have) or P2 (future).
+* It is also recommended to give requirements an identifier that will make them easier to reference later.
+-->
+
+<!--
+
+### Functional
+
+- REQ01 (P1): bla bla bla
+- REQ02 (P0): another requirement
+
+### Non-Functional
+
+- REQ03 (P1): bla bla bla
+
+-->
+
+## Proposed changes
+
+First, we would like to propose using standard types `http.Request` and `http.Response` instead of `cloud.ApiRequest` and `cloud.ApiResponse` for API handlers, encouraging reuse with any future Wing libraries and classes users write that also wish to use API middleware.
+
+Related to the above, we would like to support the SDK exporting a type `http.Handler` that is simply a type alias for `inflight (http.Request): http.Response` to reduce the amount of boilerplate in middleware code.
+
+Second, we would like to propose an `addMiddleware` method that will allow an inflight function to be automatically applied to all routes:
+
+```js
+api.addMiddleware(loggingMiddleware); // runs on all routes
+```
+
+Lastly, to support middleware being shared between Wing libraries, we would like to prioritize fixes for the two bugs mentioned in the Design section:
+- Fix issue with reassignment to variables typed as inflight functions.
+- Fix issue where inflight functions cannot be created inside preflight static methods.
+
+Here is a finished example illustrating how middleware would be shown in our docs:
+
+```js
+bring cloud;
+bring http;
+
+let api = new cloud.Api();
+
+let authMiddleware = (handler: http.Handler): http.Handler => {
+  return inflight (req) => {
+    let token = req.headers?.tryGet("x-service-token");
+    if let token = token {
+      return handler(req);
```

Comment:
How do we address the need for the user handler to obtain context from the middleware?

In this example, the user handler might want to access the token.

I am wondering if we need some arbitrary context map at the request level that middleware will be able to enrich and then the consumer will be able to access either directly (`req.ctx.get("token")`) or via a helper (`MyAuthMiddleware.getToken(req)`).